server/datastore/mysql/hosts.go

```diff
@@ -826,6 +827,42 @@ func filterHostsByStatus(now time.Time, sql string, opt fleet.HostListOptions, p
 	return sql, params
 }
 
+func filterHostsByMacOSSettingsStatus(sql string, opt fleet.HostListOptions, params []interface{}) (string, []interface{}) {
+	if opt.MacOSSettingsFilter == "" {
+		return sql, params
+	}
+
+	newSQL := ""
+	newParams := []interface{}{}
+
+	if opt.TeamFilter == nil {
+		newSQL += ` AND h.team_id IS NULL`
+	}
+
+	newSQL += ` AND EXISTS (SELECT 1 FROM host_mdm_apple_profiles hmap WHERE hmap.host_uuid = h.uuid`
+
+	switch opt.MacOSSettingsFilter {
+	case fleet.MacOSSettingsStatusFailing:
```

Comment:
I haven't checked the surrounding code, so I don't know how plausible this is, but could we use the same trick we did for the aggregations and pre-compute an aggregate of each host status and we use the number of `failed`, `pending` or `applied` profiles to determine this?

otherwise we have to execute this subquery and loop through the `host_mdm_apple_profiles` table for each `h.uuid`

alternatively, as minor optimizations, can we add `LIMIT 1` to all the clauses that look for a `failed` status? we only need one to know that the host is in a failing state

Reply:
Local performance results with 2000+ records in `hosts` (no index for `hosts.uuid`) and 4000+ records in `host_mdm_apple_profiles`:

1. Precompute aggregate status of each host (120ms): 
```sql
 SELECT
	h.*,
	hmap.*
FROM
	hosts h
	LEFT JOIN (
		SELECT
			host_uuid,
			COUNT(
				CASE WHEN status = 'applied' THEN
					1
				END) AS applied,
			COUNT(
				CASE WHEN status = 'failed' THEN
					1
				END) AS failed,
			COUNT(
				CASE WHEN status = 'pending' THEN
					1
				END) AS pending
		FROM
			host_mdm_apple_profiles
		GROUP BY
			host_uuid) hmap ON h.uuid = hmap.host_uuid
WHERE
	hmap.applied
	AND not(hmap.failed
		OR hmap.pending)
	AND h.team_id IS NULL
```
Explain:
<img width="1166" alt="Screenshot 2023-03-03 at 12 28 00 PM" src="https://user-images.githubusercontent.com/73313222/222799354-c8fdf9f0-bc13-446c-aa53-614c05457423.png">


2. Sub-select (27ms):
```sql
SELECT
	h.*
FROM
	hosts h
WHERE
	EXISTS (
		SELECT
			1
		FROM
			host_mdm_apple_profiles hmap
		WHERE
			hmap.host_uuid = h.uuid
			AND hmap.status = 'applied'
			AND NOT EXISTS (
				SELECT
					1
				FROM
					host_mdm_apple_profiles hmap2
				WHERE
					h.uuid = hmap2.host_uuid
					AND(hmap2.status = 'failed'
						OR hmap2.status = 'pending')))
			AND h.team_id IS NULL
```
Explain:
<img width="1166" alt="Screenshot 2023-03-03 at 12 28 17 PM" src="https://user-images.githubusercontent.com/73313222/222799427-04e5bb0e-83bc-40e9-8257-4c7a5598e388.png">

Reply:
Looks like the revised sub-select is performant even without the `hosts.uuid` index because the `host_mdm_apple_profiles` table is well indexed.

Reply:
ohhhh, right! you only have to compute one of the states 🤦 , thanks for checking and great use of `EXISTS`